### ISSUE TITLE
ivalue::Future then callback could hang

### ISSUE THREAD

**Issue by mrshenli** [pytorch/pytorch#58239]
It's very reproducible for me locally with the following code. After removing the `.then()` callback, it no longer hangs.  

```python
import torch
from torch.distributed import rpc
import torch.multiprocessing as mp

import os

def identity(x):
    return x

def run():
    i = 0
    while True:
        x = torch.randn(100, 100)
        futs = []
        for index in range(50):
            futs.append(
                rpc.rpc_async(
                    "server", identity, args=(x,)
                ).then(lambda x: x.wait()) # it no longer hangs after removing this then cb
            )

        torch.futures.wait_all(futs)
        print(f"== round {i}: {futs[0].wait().sum()}")
        i += 1

def run_worker(rank, world_size):
    os.environ['MASTER_ADDR'] = 'localhost'
    os.environ['MASTER_PORT'] = '29500'

    if rank == 1:
        options = rpc.TensorPipeRpcBackendOptions(
                num_worker_threads = 256,
                device_maps={"server" : {0 : 1}}
        )
        rpc.init_rpc("client", rank=rank, world_size=world_size, rpc_backend_options=options)
        run()
    else:
        rpc.init_rpc("server", rank=rank, world_size=world_size)

    rpc.shutdown()


if __name__=="__main__":
    world_size = 2
    mp.spawn(run_worker, args=(world_size, ), nprocs=world_size, join=True)

```

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @rohan-varma @jjlilley @osalpekar @jiayisuse @mrzzd @agolynski @SciPioneer @H-Huang @cbalioglu @gcramer23

**Comment by mrshenli** [comment]
cc @lw

**Comment by mrshenli** [comment]
The problem also seems to disappear after I switch to `add_done_callback`

**Comment by lw** [comment]
This turned out to be a lock inversion between the Future's mutex and the GIL. I have a (hacky) fix out for it in https://github.com/pytorch/pytorch/pull/58382.

Let me also perhaps explain how I found this out, so that next time you can diagnose this on your own. After launching the repro I attached GDB to the two processes (via `gdb --pid=<PID>`; I found the PIDs using `pgrep`). I looked at the main thread first (`thread 1` to switch to the first thread) as that's always a good place to start. I printed the stack traces (`backtrace`) and one of the processes looked very suspicious:
```
#0  __lll_lock_wait (futex=futex@entry=0x7f76f1ed2598, private=0x0) at lowlevellock.c:52
#1  0x00007f799c183c42 in __GI___pthread_mutex_lock (mutex=0x7f76f1ed2598) at ../nptl/pthread_mutex_lock.c:80
#2  0x00007f790845dee3 in __gthread_mutex_lock (__mutex=0x7f76f1ed2598) at third-party-buck/platform009/build/libgcc/include/c++/9.x/x86_64-facebook-linux/bits/gthr-default.h:749
#3  0x00007f790845de98 in std::mutex::lock (this=0x7f76f1ed2598) at third-party-buck/platform009/build/libgcc/include/c++/9.x/bits/std_mutex.h:100
#4  0x00007f79088a4eac in std::unique_lock<std::mutex>::lock (this=0x7ffcac6e2d48) at third-party-buck/platform009/build/libgcc/include/c++/9.x/bits/unique_lock.h:142
#5  0x00007f79088a4a64 in std::unique_lock<std::mutex>::unique_lock (this=0x7ffcac6e2d48, __m=...) at third-party-buck/platform009/build/libgcc/include/c++/9.x/bits/unique_lock.h:72
#6  0x00007f790889f69e in c10::ivalue::Future::addCallback(std::function<void (c10::ivalue::Future&)>) (this=0x7f76f1ed2580, callback=...) at caffe2/aten/src/ATen/core/ivalue_inl.h:477
#7  0x00007f7919902a17 in c10::collectAll (srcs=...) at caffe2/aten/src/ATen/core/ivalue.cpp:1000
#8  0x00007f7908b71882 in torch::jit::initJITBindings(_object*)::$_148::operator()(std::vector<std::shared_ptr<torch::jit::PythonFutureWrapper>, std::allocator<std::shared_ptr<torch::jit::PythonFutureWrapper> > > const&) const (this=0x7f78c5818b38, futures=...) at caffe2/torch/csrc/jit/python/init.cpp:1390
#9  0x00007f7908b715e7 in pybind11::detail::argument_loader<std::vector<std::shared_ptr<torch::jit::PythonFutureWrapper>, std::allocator<std::shared_ptr<torch::jit::PythonFutureWrapper> > > const&>::call_impl<std::shared_ptr<torch::jit::PythonFutureWrapper>, torch::jit::initJITBindings(_object*)::$_148&, 0ul, pybind11::detail::void_type>(torch::jit::initJITBindings(_object*)::$_148&, std::integer_sequence<unsigned long, 0ul>, pybind11::detail::void_type&&) && (this=0x7ffcac6e3120, f=...) at third-party-buck/platform009/build/pybind11/11419d1/include/pybind11/cast.h:2042
#10 0x00007f7908b701df in pybind11::detail::argument_loader<std::vector<std::shared_ptr<torch::jit::PythonFutureWrapper>, std::allocator<std::shared_ptr<torch::jit::PythonFutureWrapper> > > const&>::call<std::shared_ptr<torch::jit::PythonFutureWrapper>, pybind11::detail::void_type, torch::jit::initJITBindings(_object*)::$_148&> (this=0x7ffcac6e3120, f=...) at third-party-buck/platform009/build/pybind11/11419d1/include/pybind11/cast.h:2014
```
where we see `Future::addCallback` getting stuck acquiring a lock, but the only lock we acquire there is the Future's own mutex, on which we expect little contention, so there's definitely something going on. The question was who else was holding a lock on that mutex. I don't know of any other way than going through the other threads and inspecting their stack traces, which here was daunting because you configured it to use 256 worker threads. However, GDB's thread overview (`info thread`) shows the tip of the stack for all threads and their names, which allowed me to quickly rule out irrelevant threads and to see that most threads looked similar (probably all idle) whereas a few of them had a different tip and this made them "outliers". I looked into them first and indeed after 2-3 tries I found the culprit, which looked like this:
```
#0  futex_abstimed_wait_cancelable (private=<optimized out>, abstime=0x7f7826b1b1f0, clockid=<optimized out>, expected=0x0, futex_word=0x7f799c5054e8 <_PyRuntime+1224>)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:208
#1  __pthread_cond_wait_common (abstime=0x7f7826b1b1f0, clockid=<optimized out>, mutex=0x7f799c5054f0 <_PyRuntime+1232>, cond=0x7f799c5054c0 <_PyRuntime+1184>)
    at pthread_cond_wait.c:520
#2  __pthread_cond_timedwait (cond=cond@entry=0x7f799c5054c0 <_PyRuntime+1184>, mutex=mutex@entry=0x7f799c5054f0 <_PyRuntime+1232>, abstime=abstime@entry=0x7f7826b1b1f0)
    at pthread_cond_wait.c:656
#3  0x00007f799c35bfe0 in PyCOND_TIMEDWAIT (us=<optimized out>, mut=0x7f799c5054f0 <_PyRuntime+1232>, cond=0x7f799c5054c0 <_PyRuntime+1184>) at Python/condvar.h:73
#4  take_gil (ceval=ceval@entry=0x7f799c505268 <_PyRuntime+584>, tstate=tstate@entry=0x7f76f1dc6b80) at Python/ceval_gil.h:206
#5  0x00007f799c3afa45 in PyEval_AcquireThread (tstate=0x7f76f1dc6b80) at Python/ceval.c:316
#6  0x00007f790840dbef in pybind11::gil_scoped_acquire::gil_scoped_acquire (this=0x7f7826b1b360)
    at third-party-buck/platform009/build/pybind11/11419d1/include/pybind11/pybind11.h:2143
#7  0x00007f7908bfe22b in c10::ivalue::ConcretePyObjectHolder::extractTensors (this=0x7f76f1de9e40) at caffe2/torch/csrc/jit/python/python_ivalue.h:51
#8  0x00007f791990233f in c10::ivalue::Future::extractDataPtrs (value=...) at caffe2/aten/src/ATen/core/ivalue.cpp:945
#9  0x00007f79088a86fc in c10::ivalue::Future::markCompleted (this=0x7f76f1ed2580, value=..., data_ptrs=...) at caffe2/aten/src/ATen/core/ivalue_inl.h:380
```
which apparently is stuck holding the GIL (which is a red flag on its own), but earlier in the stack it was in `markCompleted`, which also acquires the Future's mutex. Hence this thread is probably the one that's preventing the main thread from advancing. After looking back at the main thread's stack trace I noticed that it contained some Python frames too, which made me suspicious that this was a lock inversion: the main thread was holding the GIL and was attempting to acquire the Future's mutex, whereas this second thread was doing the reverse.

I then quickly hacked a fix that released the GIL in the main thread and indeed the repro then worked.

**Comment by mrshenli** [comment]
closing as this is fixed by #58382
